src/libraries/System.Linq.AsyncEnumerable/src/System/Linq/Shuffle.cs

```diff
@@ -0,0 +1,59 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System.Collections.Generic;
+using System.Runtime.CompilerServices;
+using System.Threading;
+
+namespace System.Linq
+{
+    public static partial class AsyncEnumerable
+    {
+#if !NET
+        [ThreadStatic]
+        private static Random? t_random;
+#endif
+
+        /// <summary>Shuffles the order of the elements of a sequence.</summary>
+        /// <typeparam name="TSource">The type of the elements of <paramref name="source"/>.</typeparam>
+        /// <param name="source">A sequence of values to shuffle.</param>
+        /// <returns>A sequence whose elements correspond to those of the input sequence in randomized order.</returns>
+        /// <remarks>Randomization is performed using a non-cryptographically-secure random number generator.</remarks>
+        public static IAsyncEnumerable<TSource> Shuffle<TSource>(
+            this IAsyncEnumerable<TSource> source)
+        {
+            ThrowHelper.ThrowIfNull(source);
+
+            return Impl(source, default);
+
+            static async IAsyncEnumerable<TSource> Impl(
+                IAsyncEnumerable<TSource> source,
+                [EnumeratorCancellation] CancellationToken cancellationToken)
+            {
+                TSource[] array = await source.ToArrayAsync(cancellationToken).ConfigureAwait(false);
+
+#if NET
+                Random.Shared.Shuffle(array);
+#else
+                Random random = t_random ??= new Random(Environment.TickCount ^ Environment.CurrentManagedThreadId);
```

Comment:
Not a huge deal, but couldn't this be expressed as an extension method we polyfill?

Reply:
Not until we have extension properties and extension statics.

A real polyfill for it would also be much more code, as the instance returned by Shared is usable from any thread, not just the current one.